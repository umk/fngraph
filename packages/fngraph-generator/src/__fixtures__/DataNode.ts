import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
