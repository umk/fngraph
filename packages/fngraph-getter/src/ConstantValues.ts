import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
