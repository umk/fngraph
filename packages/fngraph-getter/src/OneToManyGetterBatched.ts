import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  ContextArrayGroup,
  ContextGroupInversion,
  DEFAULT_MAX_BATCH,
  DEFAULT_MAX_BUFFER,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
  PropertyRef,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'
import RecordBatchGroup, { getRecordBatchGroup } from './RecordBatchGroup'

type OneToManyGetterBatched<P extends DataRecord, R extends DataRecord> = (
  record: Array<P>,
  properties: Array<PropertyRef>,
) => Promise<Array<Array<R | Falsy>>>

export function createOneToManyBatched<P extends DataRecord, R extends DataRecord>(
  source: OneToManyGetterBatched<P, R>,
  maxBatch: number = DEFAULT_MAX_BATCH,
  maxBuffer: number = DEFAULT_MAX_BUFFER,
): GetterFactory<P, R> {
  return function (incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (let value of records) {
        context.advanceToValueGroup(value)
        if (context.groups.length > context.groupIndex) {
          const derivations = context.groups[context.groupIndex] as ContextArrayGroup<R> | undefined
          if (derivations) {
            // The cached derived record is defined only if and only
            // if the context is also defined.
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const base = getRecord(value.contexts)!
            yield* iteration(value, base, derivations as Array<R>)
          }
          context.groupIndex++
        } else {
          const groups = new Map<number, RecordBatchGroup<R>>()
          // The buffer maintains correct order of yield
          const groupIndexes: Array<number> = []
          let batchSize = 0
          for (let done: boolean | undefined = false; ; ) {
            if (!done) {
              const base = getRecord(value.contexts)
              if (!groups.has(context.groupIndex)) {
                const group = getRecordBatchGroup<R>(context, value, base)
                groups.set(context.groupIndex, group)
                if (group.include) batchSize++
              }
              groupIndexes.push(context.groupIndex)
              context.groupIndex++
            }
            if (groups.size === maxBatch || groupIndexes.length === maxBuffer || done) {
              if (batchSize > 0) {
                const g = Array.from(groups.values()).filter((g) => g.include)
                ;(
                  await source(
                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                    g.map((g) => incoming(g.base!)),
                    properties,
                  )
                )
                  .map((r) => {
                    if (!invert) return r.filter(Boolean)
                    return r.some(Boolean) ? undefined : [CONTEXT_GROUP_INVERSION]
                  })
                  .forEach(
                    (r, i) => (g[i].derived = r as Array<R | ContextGroupInversion> | undefined),
                  )
              }
              for (const groupIndex of groupIndexes) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const group = groups.get(groupIndex)!
                if (context.isCached) context.groups[groupIndex] = group.derived
                if (group.derived) {
                  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                  yield* iteration(group.value, group.base!, group.derived as Array<R>)
                }
              }
              break
            }
            ;({ value, done } = await records.next())
            if (!done) context.advanceToValueGroup(value)
          }
        }
      }
    }
  }
}

export default OneToManyGetterBatched
