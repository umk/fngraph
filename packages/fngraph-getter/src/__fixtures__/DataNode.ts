import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
