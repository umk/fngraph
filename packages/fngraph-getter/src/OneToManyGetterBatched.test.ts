import { DataRecord, DeclarationID } from '@fngraph/data'
import {
  contextAsRecord,
  createRecordGenerator,
  DataNodeSequence,
  recordAsContext,
} from '@fngraph/generator'

import { createConstant } from './ConstantValues'
import { createOneToManyBatched } from './OneToManyGetterBatched'
import { createDataNode } from './__fixtures__'

describe('createOneToManyBatched', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID
  const declarationD = 'declaration D' as DeclarationID
  it('iterates through getter records', async () => {
    const values = new Array(25).fill(0).map((_, i) => ({
      [declarationA]: i * 2 + 1,
      [declarationB]: i * 2 + 2,
    }))
    const calls: Array<number> = []
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToManyBatched((records) => {
        calls.push(records.length)
        const result = records.map((record) => {
          const valueA = record[declarationA] as number
          const valueB = record[declarationB] as number
          return [{ [declarationC]: valueA + valueB }, { [declarationC]: valueA - valueB }]
        })
        return Promise.resolve(result)
      }, 7)(contextAsRecord, recordAsContext),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB])
    const generator = createRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(calls).toMatchSnapshot()
  })
  it('permutates records with another node', async () => {
    /**
     * Must result in the following evaluation sequence:
     *
     *   (B) <- (C) <- (A)
     *
     *        - or -
     *
     *   (B) <- (A) <- (C)
     */
    const values = new Array(10).fill(0).map((_, i) => ({
      [declarationA]: i * 2 + 1,
      [declarationB]: i * 2 + 2,
    }))
    const calls: Array<number> = []
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const getter = jest.fn((records: Array<DataRecord>) => {
      calls.push(records.length)
      const result = records.map((record) => {
        const valueA = record[declarationA] as number
        const valueB = record[declarationB] as number
        return [{ [declarationC]: valueA + valueB }, { [declarationC]: valueA - valueB }]
      })
      return Promise.resolve(result)
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToManyBatched(getter, 3)(contextAsRecord, recordAsContext),
    })
    const nodeC = createDataNode({
      outgoing: [declarationD],
      getter: createConstant([{ [declarationD]: 10 }, { [declarationD]: 20 }])(
        contextAsRecord,
        recordAsContext,
      ),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB, nodeC])
    const generator = createRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(calls).toMatchSnapshot()
  })
})
