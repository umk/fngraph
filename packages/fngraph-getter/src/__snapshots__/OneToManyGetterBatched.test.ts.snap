// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOneToManyBatched iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": -1,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": -1,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": -1,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": -1,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": -1,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": -1,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": -1,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": -1,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": -1,
  },
  {
    "declaration A": 21,
    "declaration B": 22,
    "declaration C": 43,
  },
  {
    "declaration A": 21,
    "declaration B": 22,
    "declaration C": -1,
  },
  {
    "declaration A": 23,
    "declaration B": 24,
    "declaration C": 47,
  },
  {
    "declaration A": 23,
    "declaration B": 24,
    "declaration C": -1,
  },
  {
    "declaration A": 25,
    "declaration B": 26,
    "declaration C": 51,
  },
  {
    "declaration A": 25,
    "declaration B": 26,
    "declaration C": -1,
  },
  {
    "declaration A": 27,
    "declaration B": 28,
    "declaration C": 55,
  },
  {
    "declaration A": 27,
    "declaration B": 28,
    "declaration C": -1,
  },
  {
    "declaration A": 29,
    "declaration B": 30,
    "declaration C": 59,
  },
  {
    "declaration A": 29,
    "declaration B": 30,
    "declaration C": -1,
  },
  {
    "declaration A": 31,
    "declaration B": 32,
    "declaration C": 63,
  },
  {
    "declaration A": 31,
    "declaration B": 32,
    "declaration C": -1,
  },
  {
    "declaration A": 33,
    "declaration B": 34,
    "declaration C": 67,
  },
  {
    "declaration A": 33,
    "declaration B": 34,
    "declaration C": -1,
  },
  {
    "declaration A": 35,
    "declaration B": 36,
    "declaration C": 71,
  },
  {
    "declaration A": 35,
    "declaration B": 36,
    "declaration C": -1,
  },
  {
    "declaration A": 37,
    "declaration B": 38,
    "declaration C": 75,
  },
  {
    "declaration A": 37,
    "declaration B": 38,
    "declaration C": -1,
  },
  {
    "declaration A": 39,
    "declaration B": 40,
    "declaration C": 79,
  },
  {
    "declaration A": 39,
    "declaration B": 40,
    "declaration C": -1,
  },
  {
    "declaration A": 41,
    "declaration B": 42,
    "declaration C": 83,
  },
  {
    "declaration A": 41,
    "declaration B": 42,
    "declaration C": -1,
  },
  {
    "declaration A": 43,
    "declaration B": 44,
    "declaration C": 87,
  },
  {
    "declaration A": 43,
    "declaration B": 44,
    "declaration C": -1,
  },
  {
    "declaration A": 45,
    "declaration B": 46,
    "declaration C": 91,
  },
  {
    "declaration A": 45,
    "declaration B": 46,
    "declaration C": -1,
  },
  {
    "declaration A": 47,
    "declaration B": 48,
    "declaration C": 95,
  },
  {
    "declaration A": 47,
    "declaration B": 48,
    "declaration C": -1,
  },
  {
    "declaration A": 49,
    "declaration B": 50,
    "declaration C": 99,
  },
  {
    "declaration A": 49,
    "declaration B": 50,
    "declaration C": -1,
  },
]
`;

exports[`createOneToManyBatched iterates through getter records 2`] = `
[
  7,
  7,
  7,
  4,
]
`;

exports[`createOneToManyBatched permutates records with another node 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
    "declaration D": 10,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
    "declaration D": 10,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
    "declaration D": 10,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
    "declaration D": 10,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
    "declaration D": 10,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
    "declaration D": 10,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
    "declaration D": 10,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
    "declaration D": 10,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": -1,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 20,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
    "declaration D": 20,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
    "declaration D": 20,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
    "declaration D": 20,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
    "declaration D": 20,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
    "declaration D": 20,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
    "declaration D": 20,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
    "declaration D": 20,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": -1,
    "declaration D": 20,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
    "declaration D": 20,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": -1,
    "declaration D": 20,
  },
]
`;

exports[`createOneToManyBatched permutates records with another node 2`] = `
[
  3,
  3,
  3,
  1,
]
`;
